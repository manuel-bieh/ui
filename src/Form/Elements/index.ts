export { default as Checkbox } from './Checkbox';
export { default as Field } from './Field';
export { default as Input } from './Input';
export { default as Radio } from './Radio';
export { default as Submit } from './Submit';
export { default as Textarea } from './Textarea';
