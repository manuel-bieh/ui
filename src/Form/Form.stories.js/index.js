import React from 'react';
import { storiesOf } from '@storybook/react';
import Simple from './Simple';
import AsyncValidation from './AsyncValidation';
import CheckboxTypes from './CheckboxTypes';
import Submission from './Submission';

storiesOf(Group.Components('Form'), module)
    .add('Simple form', () => <Simple />)
    .add('Checkbox types', () => <CheckboxTypes />)
    .add('Async validation', () => <AsyncValidation />)
    .add('Submission', () => <Submission />);
