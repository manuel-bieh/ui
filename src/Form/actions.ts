type Values = {
    [key: string]: any;
};

export const setValue = (name: string, value: any) => ({
    type: 'SET_VALUE',
    payload: {
        name,
        value,
    },
});

export const setValues = (values: Values) => ({
    type: 'SET_VALUES',
    payload: values,
});

export const validationInProgress = (isValidating: boolean) => ({
    type: 'SET_VALIDATION_STATUS',
    payload: isValidating,
});

export const resetError = (fieldName: string) => ({
    type: 'RESET_FIELD_ERROR',
    payload: fieldName,
});

export const setErrors = (errors: any) => ({
    type: 'SET_ERRORS',
    payload: errors,
});

export const setExternalErrors = (errors: any) => ({
    type: 'SET_EXTERNAL_ERRORS',
    payload: errors,
});

export const setSubmissionStatus = (isSubmitting: boolean) => ({
    type: 'SET_SUBMISSION_STATUS',
    payload: isSubmitting,
});
