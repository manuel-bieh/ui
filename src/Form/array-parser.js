/* eslint-disable no-unused-vars,no-var,no-useless-escape,no-redeclare */
const value = 'placeholder';
const name = 'users:test[admins][superadmins][]';
// var name = 'users-test[admins][superadmins][]';
// var name = 'users_test[admins][superadmins][]';
// var name = 'users[admins][superadmins][]';
// var [fieldName, arrayParams = ''] = name.split(/^([^\[\]]*)/).filter(Boolean);
// arrayParams.split(/(\[[\w-:_]*\])/).filter(Boolean);

// var name = 'users[admins][superadmins]';
// var name = 'users[admins][superadmins]';
var [fieldName, arrayParams = ''] = name.split(/^([^\[\]]*)/).filter(Boolean);
var arrayParams = arrayParams.split(/(\[[^(\[|\])]*\])/).filter(Boolean);
const obj = {};
const reduced = [fieldName, ...arrayParams].reduce(
    (acc, fieldName, index, array) => {
        const [obj, last] = acc;
        const next = last ? last : obj;
        const [, cleanName] = fieldName.match(/^\[(.*)\]$/) || [null, fieldName];
        if (fieldName !== '[]') {
            next[cleanName] = array[index + 1] === '[]' && array.length - 1 === index + 1 ? [] : {};
        }

        const isLastItem = array.length - 1 === index;
        return isLastItem ? obj : [obj, next[cleanName]];
    },
    [{}]
);

console.log(reduced);

// 'a.b.c'.split('.').reduce((acc, name, index, array) => {
//   const [obj, last] = acc;
//   const next = last ? last : obj;
//   next[name] = {};
//   return array.length - 1 === index ? obj : [obj, next[name]];
// }, [{}])

//     if (item !== '[]') {
//         next[item] = array[index + 1] === '[]' && array.length - 1 === index + 1 ? [] : {};
//         return next[item];
//     }
//     return next;
// }, obj);
// console.log(obj);
// var name = 'users[admins][superadmins]';
// var [fieldName, arrayParams = ''] = name.split(/^([^\[\]]*)/).filter(Boolean);
// console.log(fieldName);
// var arrayParams = arrayParams.split(/(\[[^(\[|\])]*\])/).filter(Boolean);
// var obj = {};
// var reduced = [fieldName, ...arrayParams].reduce((next, item, index, array) => {
//     if (item !== '[]') {
//         next[item] = array[index + 1] === '[]' && array.length - 1 === index + 1 ? [] : {};
//         return next[item];
//     }
//     return next;
// }, obj);
// console.log(obj);

// var obj = {
//   [fieldName]: arrayParams.length === 0 ? null : arrayParams.reduce((next, item) => {
//       if (item !== '[]') {
//           next[item] = {};
//       }
//       return next;
//   }, obj[fieldName])
// }

// var name = 'users';
// var values = {
//     users: value,
// };

// var name = 'users[]';
// var values = {
//     users: [value],
// };

// var name = 'users[admins][superadmins][]';
// var values = {
//     users: {
//         admins: {
//             superadmins: [],
//         },
//     },
// };

// var name = 'users[admins][superadmins][1]';
// var values = {
//     users: {
//         admins: {
//             superadmins: { 1: value },
//         },
//     },
// };
