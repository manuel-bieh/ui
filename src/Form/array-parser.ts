// const value = 'placeholder';
// var name = 'users:test[admins][superadmins][]';
// var name = 'users-test[admins][superadmins][]';
// var name = 'users_test[admins][superadmins][]';
// var name = 'users[admins][superadmins][]';
// var [fieldName, arrayParams = ''] = name.split(/^([^\[\]]*)/).filter(Boolean);
// arrayParams.split(/(\[[\w-:_]*\])/).filter(Boolean);

// var name = 'users[admins][superadmins]';
// var name = 'users[admins][superadmins]';
// var [fieldName, arrayParams = ''] = name.split(/^([^\[\]]*)/).filter(Boolean);
// var arrayParams = arrayParams.split(/(\[[^(\[|\])]*\])/).filter(Boolean);
// const obj = {};
// const reduxed = [fieldName, ...arrayParams].reduce(
//     (acc, fieldName, index, array) => {
//         const [obj, last] = acc;
//         const next = last ? last : obj;
//         const [, cleanName] = fieldName.match(/^\[(.*)\]$/) || [
//             null,
//             fieldName,
//         ];
//         if (fieldName !== '[]') {
//             next[cleanName] =
//                 array[index + 1] === '[]' && array.length - 1 === index + 1
//                     ? []
//                     : {};
//         }
//         return array.length - 1 === index ? obj : [obj, next[cleanName]];
//     },
//     [{}]
// );

// console.log(reduxed);

// 'a.b.c'.split('.').reduce((acc, name, index, array) => {
//   const [obj, last] = acc;
//   const next = last ? last : obj;
//   next[name] = {};
//   return array.length - 1 === index ? obj : [obj, next[name]];
// }, [{}])

//     if (item !== '[]') {
//         next[item] = array[index + 1] === '[]' && array.length - 1 === index + 1 ? [] : {};
//         return next[item];
//     }
//     return next;
// }, obj);
// console.log(obj);
// var name = 'users[admins][superadmins]';
// var [fieldName, arrayParams = ''] = name.split(/^([^\[\]]*)/).filter(Boolean);
// console.log(fieldName);
// var arrayParams = arrayParams.split(/(\[[^(\[|\])]*\])/).filter(Boolean);
// var obj = {};
// var reduxed = [fieldName, ...arrayParams].reduce((next, item, index, array) => {
//     if (item !== '[]') {
//         next[item] = array[index + 1] === '[]' && array.length - 1 === index + 1 ? [] : {};
//         return next[item];
//     }
//     return next;
// }, obj);
// console.log(obj);

// var obj = {
//   [fieldName]: arrayParams.length === 0 ? null : arrayParams.reduce((next, item) => {
//       if (item !== '[]') {
//           next[item] = {};
//       }
//       return next;
//   }, obj[fieldName])
// }

// var name = 'users';
// var values = {
//     users: value,
// };

// var name = 'users[]';
// var values = {
//     users: [value],
// };

// var name = 'users[admins][superadmins][]';
// var values = {
//     users: {
//         admins: {
//             superadmins: [],
//         },
//     },
// };

// var name = 'users[admins][superadmins][1]';
// var values = {
//     users: {
//         admins: {
//             superadmins: { 1: value },
//         },
//     },
// };
