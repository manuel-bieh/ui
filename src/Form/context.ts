import React from 'react';

const Context = React.createContext<any>(null);

export const Consumer = Context.Consumer;
export const Provider = Context.Provider;

export default Context;
