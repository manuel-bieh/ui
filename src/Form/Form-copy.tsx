import * as React from 'react';
import { Provider } from './context';
import formReducer from './reducer';
import * as actions from './actions';

type FormValuesT = {
    [key: string]: any;
};

type PropsT = {
    children: any;
    id?: string;
    initialValues?: FormValuesT;
    onUnmount?: (data: { id: string; values: any }) => any;
    onMount?: (datay: { id: string; values: any }) => any;
    onUpdate?: (data: { id: string; values: any }) => any;
    onSubmit: (data: { form: any; id: any; values: any }) => any;
    validation?: (any) => any;
};

type StateT = {
    [key: string]: any;
    errors?: any;
    isSubmitting: boolean;
    isValidating: boolean;
    values: FormValuesT;
};

type FormContextT = {
    errors?: any;
    externalErrors?: any;
    form: null | any;
    id?: string;
    isInvalid: (fieldName: string) => boolean;
    isSubmitting: boolean;
    isValidating: boolean;
    onChange: (e: React.ChangeEvent) => any;
    resetValues: () => void;
    resetError: (fieldName: string) => void;
    submit: () => any;
    values: FormValuesT;
};

type ValuesObject = {
    [key: string]: any;
};

class FormClass extends React.PureComponent<PropsT, StateT> {
    state = {
        errors: null,
        isSubmitting: false,
        isValidating: false,
        externalErrors: null,
        values: this.props.initialValues || {},
    };

    componentDidMount() {
        this.setState(
            () => ({
                values: this.props.initialValues || {},
            }),
            this.onMount
        );
    }

    componentDidUpdate(prevProps: PropsT, prevState: StateT) {
        const { id, onUpdate } = this.props;
        const { values } = this.state;
        if (values !== prevState.values && typeof onUpdate === 'function') {
            onUpdate({ id, values });
        }
    }

    componentWillUnmount() {
        this.onUnmount();
    }

    form: { current: null | HTMLFormElement } = React.createRef();

    isInvalid = (fieldName: string) =>
        (this.state.errors && this.state.errors[fieldName]) || false;

    resetError = (fieldName: string) => {
        this.setState((state) => {
            const errors = state.errors ? { ...state.errors } : {};
            delete errors[fieldName];
            return {
                errors: Object.keys(errors || {}).length === 0 ? null : errors,
            };
        });
    };

    setErrors = (errors: { [string]: string }) => {
        this.setState(() => ({ errors }));
    };

    setValue = (property: string, value: any) => {
        this.setState(({ values }) => ({
            values: {
                ...values,
                [property]: value,
            },
        }));
    };

    onMount = () => {
        const { id, onMount } = this.props;
        const { values } = this.state;

        if (typeof onMount === 'function') {
            onMount({ id, values });
        }
    };

    onUnmount = () => {
        const { id, onUnmount } = this.props;
        const { values } = this.state;

        if (typeof onUnmount === 'function') {
            onUnmount({ id, values });
        }
    };

    onChange = (e: React.ChangeEvent) => {
        e.persist();
        const {
            target: { checked, name, type, value },
        } = e;

        if (!name && process.env.NODE_ENV === 'development') {
            console.warn(
                `Change event with value '${value}' happened for an element without name prop. This is probably by mistake.`
            );
        }

        // TODO: figure out if this is really such a good idea
        this.resetError(name);

        // TODO: make this more flexible (see the work that has been done in array-parser.js for more)
        if (type === 'checkbox') {
            if (!value || value === 'on') {
                return this.setState((state) => ({
                    values: {
                        ...state.values,
                        [name]: checked ? 'on' : undefined,
                    },
                }));
            }

            return this.setState((state) => ({
                values: {
                    ...state.values,
                    [name]: (checked
                        ? (Array.isArray(state.values[name])
                              ? state.values[name]
                              : []
                          ).concat(value)
                        : (Array.isArray(state.values[name])
                              ? state.values[name]
                              : []
                          ).filter((item) => item !== value)
                    ).reduce(
                        (acc, item) => (item ? (acc || []).concat(item) : acc),
                        undefined
                    ),
                },
            }));
        }

        // TODO: untested as of 2019-02-01
        this.setValue(name, value);

        // this.setState((state) => ({
        //     values: {
        //         ...state.values,
        //         [name]: value,
        //     },
        // }));
    };

    onSubmit = (e: SyntheticEvent<HTMLFormElement>) => {
        e.preventDefault();
        this.submit();
    };

    validate = async () => {
        const { validation } = this.props;
        const { values } = this.state;

        if (typeof validation !== 'function') {
            return true;
        }

        this.setState(() => ({
            isValidating: true,
            errors: null,
            externalErrors: null,
        }));
        try {
            const errors = await validation({ values });
            if (!errors || Object.keys(errors).length === 0) {
                return true;
            }
            this.setErrors(errors);
            return false;
        } catch (error) {
            console.error('error:', error);
            this.setState(() => ({
                externalErrors: error.message,
            }));
        } finally {
            this.setState(() => ({ isValidating: false }));
        }
    };

    submit = async () => {
        const { id, onSubmit } = this.props;
        const { isSubmitting, values } = this.state;
        const form = this.form.current;

        if (typeof onSubmit !== 'function' || isSubmitting === true) {
            return;
        }

        const isValid = await this.validate();

        if (!isValid) {
            return;
        }

        this.setState(() => ({ isSubmitting: true }));
        try {
            await onSubmit({ form, id, resetValues: this.resetValues, values });
        } catch (error) {
            console.error('error', error);
            this.setState(() => ({
                externalErrors: error.message,
            }));
        } finally {
            this.setState(() => ({ isSubmitting: false }));
        }
    };

    resetValues = (forceEmpty) => {
        // forceEmpty makes sure all values are completely nulled instead of set to their initialValues
        this.setState(() => ({
            values: forceEmpty ? {} : this.props.initialValues,
        }));
    };

    render() {
        const { children, id } = this.props;
        const {
            errors,
            isSubmitting,
            isValidating,
            externalErrors,
            values,
        } = this.state;

        const context: FormContextT = {
            form: this.form.current,
            isInvalid: this.isInvalid,
            onChange: this.onChange,
            resetValues: this.resetValues,
            resetError: this.resetError,
            setValue: this.setValue,
            submit: this.submit,
            errors,
            externalErrors,
            id,
            isSubmitting,
            isValidating,
            values,
        };

        return (
            <Provider value={context}>
                <form
                    id={this.props.id}
                    ref={this.form}
                    onSubmit={this.onSubmit}
                >
                    {typeof children === 'function'
                        ? children(context)
                        : children}
                </form>
            </Provider>
        );
    }
}

// export default React.memo(Form);
export default FormClass;
