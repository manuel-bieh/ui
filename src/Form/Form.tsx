import React, {
    useCallback,
    useEffect,
    useMemo,
    useReducer,
    useRef,
} from 'react';
import { Provider } from './context';
import formReducer from './reducer';
import * as actions from './actions';

type ValuesObject = {
    [key: string]: any;
};

type ErrorsObject = {
    [key: string]: undefined | string | boolean;
};

type Props = {
    id?: string;
    initialValues?: ValuesObject;
    onMount?: (values: ValuesObject) => any;
    onUnmount?: (values: ValuesObject) => any;
    onUpdate?: (any: any) => any;
    onSubmit: (TEMP: any) => void;
    validation?: ({
        values,
    }: {
        values: ValuesObject;
    }) => Promise<ErrorsObject>;
};

const Form: React.FC<Props> = ({
    children,
    id,
    initialValues,
    onMount,
    onUnmount,
    onUpdate,
    onSubmit,
    validation,
}) => {
    const form = useRef<any>(null);
    const isMounted = useRef(null);

    const initialState = {
        values: initialValues || {},
    };

    const [state, dispatch] = useReducer(formReducer, initialState);

    const setValue = useCallback((property: string, value: any) => {
        dispatch(actions.setValue(property, value));
    }, []);

    const setValues = useCallback((values) => {
        dispatch(actions.setValues(values));
    }, []);

    const setErrors = useCallback((errors: ErrorsObject) => {
        dispatch(actions.setErrors(errors));
    }, []);

    const setExternalErrors = useCallback((errors: ErrorsObject) => {
        dispatch(actions.setExternalErrors(errors));
    }, []);

    const setSubmissionStatus = useCallback((isSubmitting: boolean) => {
        dispatch(actions.setSubmissionStatus(isSubmitting));
    }, []);

    const validationInProgress = useCallback((isValidating: boolean) => {
        dispatch(actions.validationInProgress(isValidating));
    }, []);

    const resetError = useCallback((fieldName: string) => {
        dispatch(actions.resetError(fieldName));
    }, []);

    const resetValues = useCallback(
        (forceEmpty) => {
            // forceEmpty makes sure all values are completely nulled instead of being set to their initialValues
            setValues(forceEmpty ? {} : initialValues);
        },
        [initialValues, setValues]
    );

    const isInvalid = useCallback(
        (fieldName: string) =>
            (state.errors && state.errors[fieldName]) || false,
        [state.errors]
    );

    const validate = useCallback(async () => {
        if (typeof validation !== 'function') {
            return true;
        }

        validationInProgress(true);
        try {
            console.log('VAL');
            const errors = await validation({ values: state.values });
            console.log('VAL2');
            console.log({ errors });
            if (!errors || Object.keys(errors).length === 0) {
                return true;
            }

            setErrors(errors);
            return false;
        } catch (error) {
            setExternalErrors(error.message);
            return false;
        } finally {
            validationInProgress(false);
        }
    }, [
        setErrors,
        setExternalErrors,
        state.values,
        validation,
        validationInProgress,
    ]);

    const submit = useCallback(async () => {
        if (typeof onSubmit !== 'function' || state.isSubmitting === true) {
            return;
        }
        const isValid = await validate();

        if (!isValid) {
            return;
        }

        setSubmissionStatus(true);
        try {
            await onSubmit({
                errors: state.errors,
                externalErrors: state.externalErrors,
                form,
                id,
                initialValues,
                resetValues,
                setValue,
                setValues,
                values: state.values,
            });
        } catch (error) {
            console.error('error', error);
            setExternalErrors(error.message);
        } finally {
            setSubmissionStatus(false);
        }
    }, [
        id,
        initialValues,
        onSubmit,
        resetValues,
        setExternalErrors,
        setSubmissionStatus,
        setValue,
        setValues,
        state.errors,
        state.externalErrors,
        state.isSubmitting,
        state.values,
        validate,
    ]);

    const submitHandler = useCallback(
        (e: React.FormEvent<HTMLFormElement>) => {
            e.preventDefault();
            console.log('SUBMITTED');
            submit();
        },
        [submit]
    );

    const onChange = useCallback(
        (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
            e.persist();
            const { name, type, value } = e.target;

            if (!name && process.env.NODE_ENV === 'development') {
                console.warn(
                    `Change event with value '${value}' happened for an element without name prop. This is probably by mistake.`
                );
            }

            // TODO: figure out if this is really such a good idea
            resetError(name);

            // TODO: make this more flexible (see the work that has been done in array-parser.js for more)
            if (e.target instanceof HTMLInputElement && type === 'checkbox') {
                const { checked } = e.target;
                const values = state.values;

                if (!values[name] && checked) {
                    // field name does not yet exist. We're setting it here
                    return setValue(name, value || 'on');
                }

                if (
                    values[name] &&
                    (typeof values[name] === 'string' ||
                        typeof values[name] === 'boolean')
                ) {
                    if (checked) {
                        // field name exists already. convert to array and add old and new value to it
                        return setValue(
                            name,
                            [values[name]].concat(value || 'on')
                        );
                    }
                    return setValue(name, undefined);
                }

                if (Array.isArray(values[name])) {
                    if (checked) {
                        const addCheckedValue = (values: any) =>
                            (Array.isArray(values[name])
                                ? values[name]
                                : []
                            ).concat(value || 'on');

                        return setValue(name, addCheckedValue(values));
                    }
                    const removeUncheckedValue = (values: any) =>
                        (Array.isArray(values[name])
                            ? values[name]
                            : []
                        ).filter((item: any) => item !== value);

                    const nextValues = removeUncheckedValue(values);

                    return setValue(
                        name,
                        nextValues.length === 1 ? nextValues[0] : nextValues
                    );
                }
            }

            return setValue(name, value);
        },
        [resetError, setValue, state.values]
    );

    // set isMounted on unmount _before_ the onUnmount cleanup effect function is run
    useEffect(() => () => (isMounted.current = false), []);

    // run each onMount/onUnmount only once per component lifetime
    useEffect(() => {
        const params = {
            errors: state.errors,
            externalErrors: state.externalErrors,
            form,
            id,
            initialValues,
            resetValues,
            setValue,
            setValues,
            values: state.values,
        };

        if (typeof onMount === 'function' && isMounted.current === null) {
            onMount(params);
        }

        return () => {
            if (
                typeof onUnmount === 'function' &&
                isMounted.current === false
            ) {
                onUnmount(params);
            }
        };
    }, [
        id,
        initialValues,
        onMount,
        onUnmount,
        resetValues,
        setValue,
        setValues,
        state.errors,
        state.externalErrors,
        state.values,
    ]);

    // set isMounted to true on mount _after_ the onMount effect function is run
    useEffect(() => {
        isMounted.current = true;
    }, []);

    useEffect(() => {
        if (typeof onUpdate === 'function') {
            onUpdate({
                errors: state.errors,
                externalErrors: state.externalErrors,
                form,
                id,
                initialValues,
                resetValues,
                setValue,
                setValues,
                values: state.values,
            });
        }
    }, [
        id,
        initialValues,
        onUpdate,
        resetValues,
        setValue,
        setValues,
        state.errors,
        state.externalErrors,
        state.values,
    ]);

    const contextValue = useMemo(
        () => ({
            id,
            isInvalid,
            errors: state.errors,
            externalErrors: state.externalErrors,
            // form: form.current,
            initialValues,
            isSubmitting: state.isSubmitting,
            isValidating: state.isValidating,
            onChange,
            resetError,
            resetValues,
            setValue,
            submit,
            validate,
            values: state.values,
        }),
        [
            id,
            initialValues,
            isInvalid,
            onChange,
            resetError,
            resetValues,
            setValue,
            state.errors,
            state.externalErrors,
            state.isSubmitting,
            state.isValidating,
            state.values,
            submit,
            validate,
        ]
    );

    return (
        <Provider value={contextValue}>
            <form id={id} ref={form} onSubmit={submitHandler}>
                {children}
            </form>
        </Provider>
    );
};

export default Form;
