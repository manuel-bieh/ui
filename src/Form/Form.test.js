import React from 'react';
import { render } from '@testing-library/react';
import Form from './Form';

describe('Form', () => {
    it('renders a form', () => {
        const form = render(<Form>Hallo</Form>);
        expect(form.container.firstChild).toMatchSnapshot();
    });
});
