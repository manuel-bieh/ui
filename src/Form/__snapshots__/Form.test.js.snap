// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders a form 1`] = `
<form>
  Hallo
</form>
`;
